File: c/removeNthFromEnd.cpp
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        if (head->next == nullptr) {
            return nullptr;
        }

        int k = 0;
        ListNode *curr = head;
        ListNode *toUpdate = head;
        while (curr->next != nullptr) {
            curr = curr->next;
            if (k >= n) {
                toUpdate = toUpdate->next;
            }
            k++;
        }

        if (k == n - 1) {
            return head->next;
        }

        toUpdate->next = toUpdate->next->next;
        return head;
    }

    void printList(ListNode *head) {
        if (head == nullptr) {
            cout << endl;
        } else {
            cout << head->val << ",";
            printList(head->next);
        }
    }
};

int main() {
    Solution solution;
    ListNode *head = new ListNode(
        1, new ListNode(
               2, new ListNode(3, new ListNode(4, new ListNode(5, nullptr)))));
    ListNode *result = solution.removeNthFromEnd(head, 1);

    solution.printList(result);
    return 0;
}

File: rust/src/bin/remove_nth_from_end.rs
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/
// 2024/03/03

mod list_node;
use crate::list_node::ListNode;

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode::new(0)));
        dummy.as_mut().unwrap().next = head;

        let mut first = &mut dummy as *mut Option<Box<ListNode>>;
        let mut second = &mut dummy as *mut Option<Box<ListNode>>;

        unsafe {
            for _ in 0..=n {
                first = &mut (*first).as_mut().unwrap().next;
            }

            while !(*first).is_none() {
                first = &mut (*first).as_mut().unwrap().next;
                second = &mut (*second).as_mut().unwrap().next;
            }

            let tmp = &mut (*second).as_mut().unwrap().next;
            (*second).as_mut().unwrap().next = (*tmp).as_mut().unwrap().next.take();
        }

        dummy.unwrap().next
    }
}

pub struct Solution {}

fn main() {
    println!(
        "{:?}",
        Solution::remove_nth_from_end(ListNode::from_array(&[1, 2, 3, 4, 5]), 2)
    ); // [1,2,3,5]

    println!(
        "{:?}",
        Solution::remove_nth_from_end(ListNode::from_array(&[1]), 1)
    ); // []

    println!(
        "{:?}",
        Solution::remove_nth_from_end(ListNode::from_array(&[1, 2]), 1)
    ); // [1]
}
